internal/oci/container.go

```diff
@@ -527,17 +532,24 @@ func (c *Container) pid() (int, error) {
 // This is the simplest way to verify we are operating on the container
 // process, and haven't run into PID wrap.
 func (c *Container) verifyPid() error {
```

Comment:
There is a thing now with this function since it understands the process state. We might be baling out on things too early/eagerly, for example:

https://github.com/cri-o/cri-o/blob/5c111c94c50a633edf10e89a3c5969c5fe55edfe/internal/oci/runtime_oci.go#L805-L809

We wouldn't attempt to deliver SIGTERM or SIGKILL should the process be in the undesired state. However, this state might be brief, or the process might still be killable.

What I am saying, we need to make sure the old flow needs to be preserved/attempted at least once, I suppose.

Reply:
oh shoot you're right. yeah we need to have two different functions. One to check for Living pre-SIGKILL, one for LivingOrSleeping for after. If we filter "D" state out for all processes we check for living, the process can flip back to R and we would totally skip the kill
/hold

Reply:
We can have two functions or pass a simple boolean to `Living()` and then to `verifyPid()` to toggle the process status check so it can be skipped on demand, for example:

```go
func (c *Container) verifyPid(checkState bool) error {
	data, err := getPidData(c.state.InitPid)
	if err != nil {
		return err
	}

	if data.StartTime != c.state.InitStartTime {
		return fmt.Errorf(
			"PID %d is running but has start time of %s, whereas the saved start time is %s. PID wrap may have occurred",
			c.state.InitPid, data.StartTime, c.state.InitStartTime,
		)
	}

        // Don't check the process (PID) state.
	if !checkState {
                return nil
        }

        // Skip states which are typically considered as invalid.
        switch data.State {
	case "D", "X", "Z":
		return fmt.Errorf("PID %d has an invalid state: %s", c.state.InitPid, data.State)
	}

	return nil
}
```

Reply:
Alternatively, `verifyPid()` would be split into `verifyPIDWrap()` and `verifyPIDState()`, and used accordingly.

Reply:
I suppose, we _always_ want to check for a PID wrap condition. Which might render the idea of having two separate helpers moot.